src/iter.rs

```diff
@@ -0,0 +1,685 @@
+pub struct TimestampedTracePackets {
+    /// Timestamp of [Self::packets] and [Self::malformed_packets]
+    pub timestamp: Timestamp,
+    pub packets: Vec<TracePacket>,
+    pub malformed_packets: Vec<MalformedPacket>,
+
+    /// The number of [TracePacket]s consumed to generate this
+    /// structure.
+    pub consumed_packets: usize,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Timestamp {
+    pub ts: DateTime,
+
+    /// In what manner this timestamp relates to the associated data
+    /// packets.
+    pub data_relation: TimestampDataRelation,
+}
+
+impl PartialOrd for Timestamp {
+    /// Sorts [Timestamp]s based on [Timestamp::ts].
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.ts.cmp(&other.ts))
+    }
+}
+
+pub struct Timestamps<'a, R>
+where
+    R: Read,
+{
+    decoder: &'a mut Decoder<R>,
+    options: TimestampConfiguration,
+    current_baseline: DateTime,
+}
+
+struct Gts {
+    pub lower: Option<u64>,
+    pub upper: Option<u64>,
+}
+impl Gts {
+    const GTS2_SHIFT: usize = 26; // see (Appendix D4.2.5).
+
+    pub fn replace_lower(&mut self, new: u64) {
+        self.lower = match self.lower {
+            None => Some(new),
+            Some(old) => {
+                let shift = 64 - new.leading_zeros();
+                Some(((old >> shift) << shift) | new)
+            }
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.lower = None;
+        self.upper = None;
+    }
+
+    pub fn merge(&self) -> Option<u64> {
+        if let (Some(lower), Some(upper)) = (self.lower, self.upper) {
+            Some((upper << Self::GTS2_SHIFT) | lower)
+        } else {
+            None
+        }
+    }
+
+    #[cfg(test)]
+    pub fn from_two(gts1: TracePacket, gts2: TracePacket) -> Self {
+        match (gts1, gts2) {
+            (
+                TracePacket::GlobalTimestamp1 { ts: lower, .. },
+                TracePacket::GlobalTimestamp2 { ts: upper },
+            ) => Self {
+                lower: Some(lower),
+                upper: Some(upper),
+            },
+            _ => unreachable!(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn from_one(gts: TracePacket) -> Self {
+        match gts {
+            TracePacket::GlobalTimestamp1 { ts: lower, .. } => Self {
+                lower: Some(lower),
+                upper: None,
+            },
+            TracePacket::GlobalTimestamp2 { ts: upper, .. } => Self {
+                lower: None,
+                upper: Some(upper),
+            },
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl<'a, R> Timestamps<'a, R>
+where
+    R: Read,
+{
+    pub(super) fn new(decoder: &'a mut Decoder<R>, options: TimestampConfiguration) -> Self {
+        assert!(
+            options.lts_prescaler != LocalTimestampOptions::Disabled,
+            "local timestamps must be enabled"
```

Comment:
The alternative is to return `Option<Self>`, but I feel like that would just add noise to the API.

Reply:
Perhaps we can statically check for `Disabled` and emit a compiler error?

Reply:
`LocalTimestampOptions::Disabled` now yield `unimplemented!()` instead.